server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -1223,4 +1160,97 @@ static class PipelineHolder {
+            return Optional.empty();
+        }
+
+        final Settings settings = indexMetadata.getSettings();
+        return Optional.of(new Pipelines(IndexSettings.DEFAULT_PIPELINE.get(settings), IndexSettings.FINAL_PIPELINE.get(settings)));
+    }
+
+    private static Optional<Pipelines> resolvePipelinesFromIndexTemplates(IndexRequest indexRequest, Metadata metadata) {
+        if (indexRequest.index() == null) {
+            return Optional.empty();
+        }
+
+        // the index does not exist yet (and this is a valid request), so match index
+        // templates to look for pipelines in either a matching V2 template (which takes
+        // precedence), or if a V2 template does not match, any V1 templates
+        String v2Template = MetadataIndexTemplateService.findV2Template(metadata, indexRequest.index(), false);
+        if (v2Template != null) {
+            final Settings settings = MetadataIndexTemplateService.resolveSettings(metadata, v2Template);
+            return Optional.of(new Pipelines(IndexSettings.DEFAULT_PIPELINE.get(settings), IndexSettings.FINAL_PIPELINE.get(settings)));
+        }
+
+        String defaultPipeline = null;
+        String finalPipeline = null;
+        List<IndexTemplateMetadata> templates = MetadataIndexTemplateService.findV1Templates(metadata, indexRequest.index(), null);
+        // order of templates are the highest order first
+        for (final IndexTemplateMetadata template : templates) {
+            final Settings settings = template.settings();
+
+            // note: the exists/get trickiness here is because we explicitly *don't* want the default value
+            // of the settings -- a non-null value would terminate the search too soon
+            if (defaultPipeline == null && IndexSettings.DEFAULT_PIPELINE.exists(settings)) {
+                defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(settings);
+                // we can not break in case a lower-order template has a final pipeline that we need to collect
+            }
+            if (finalPipeline == null && IndexSettings.FINAL_PIPELINE.exists(settings)) {
+                finalPipeline = IndexSettings.FINAL_PIPELINE.get(settings);
+                // we can not break in case a lower-order template has a default pipeline that we need to collect
+            }
+            if (defaultPipeline != null && finalPipeline != null) {
+                // we can break if we have already collected a default and final pipeline
+                break;
+            }
+        }
+
+        // having exhausted the search, if nothing was found, then use the default noop pipeline names
+        defaultPipeline = Objects.requireNonNullElse(defaultPipeline, NOOP_PIPELINE_NAME);
+        finalPipeline = Objects.requireNonNullElse(finalPipeline, NOOP_PIPELINE_NAME);
+
+        return Optional.of(new Pipelines(defaultPipeline, finalPipeline));
+    }
+
+    public static boolean hasPipeline(IndexRequest indexRequest) {
```

Comment:
It's probably wrong to ever invoke this against an IndexRequest where `indexRequest.isPipelineResolved` is `false`, right? Given that, I'm inclined to add an assert for that. That'll probably require that you rewrite your tests (slightly).

Or, if it's not wrong (but I really do think it is!) then your tests need to cover null cases, too, but I really don't like that...

Along the same lines, if `indexRequest.isPipelineResolved` is asserted to be `true`, then it must never be the case that the pipeline fields are themselves null, which although the code is indeed null safe already might be worth also asserting (maybe?).

Given those asserts, it's probably logical to add a docstring.